realhf/impl/model/nn/real_llm_base.py

```diff
@@ -482,6 +496,17 @@ def real_model_tblock_param_keys(
             f"{idx + 1}.mlp.up_proj.weight",
             f"{idx + 1}.mlp.down_proj.weight",
         ]
+    elif config.mlp_type == "moe":
+        num_experts = config.moe.num_experts
+        keys += [
+            f"{idx + 1}.mlp.router.weight",
+        ]
+        for j in range(num_experts):
+            keys += [
+                f"{idx + 1}.mlp.experts.local_experts.{j}.gate_proj.weight",
+                f"{idx + 1}.mlp.experts.local_experts.{j}.up_proj.weight",
+                f"{idx + 1}.mlp.experts.local_experts.{j}.down_proj.weight",
+            ]
```

Comment:
Does the key have the same order as in the state dict? I think the actual order should be

```python
for j in range(num_experts):
    keys += [f"{idx+1}.router", f"{idx+1}.w1", f"{idx+1}.w2", f"{idx+1}.w3"]
```

If the order is not the same, it will be a bug in parameter reallocation.

Reply:
Fixed

Reply:
My previous comment was not correct. The order has been checked to be coorect, right?

Reply:
`['1.attn.c_attn.ln.weight', '1.attn.c_attn.q_attn.weight', '1.attn.c_attn.k_attn.weight', '1.attn.c_attn.v_attn.weight', '1.attn.c_proj.weight', '1.mlp.ln.weight', '1.mlp.router.weight', '1.mlp.experts.local_experts.0.gate_proj.weight', '1.mlp.experts.local_experts.0.up_proj.weight', '1.mlp.experts.local_experts.0.down_proj.weight', '1.mlp.experts.local_experts.1.gate_proj.weight', '1.mlp.experts.local_experts.1.up_proj.weight', '1.mlp.experts.local_experts.1.down_proj.weight', '1.mlp.experts.local_experts.2.gate_proj.weight', '1.mlp.experts.local_experts.2.up_proj.weight', '1.mlp.experts.local_experts.2.down_proj.weight', '1.mlp.experts.local_experts.3.gate_proj.weight', '1.mlp.experts.local_experts.3.up_proj.weight', '1.mlp.experts.local_experts.3.down_proj.weight']`

I double checked, the original one should be correct. Reverted.